# Learning Pandas here

In [140]:
import pandas as pd

## Loading CSV data

Pandas gives a great way to consume data from different format like CSV, JSON etc.

In [141]:
df = pd.read_csv('../../data/stackoverflow/survey_results_public.csv', index_col='Respondent')
df.head()

MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code      Yes   
2                      I am a student who is learning to code       No   
3           I am not primarily a developer, but I write co...      Yes   
4                              I am a developer by profession       No   
5                              I am a developer by profession      Yes   

                           OpenSourcer  \
Respondent                               
1                                Never   
2              Less than once per year   
3                                Never   
4                                Never   
5           Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   
5           OSS is, on average, of HIGHER quality than pro...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   
3                               Employed full-time                Thailand   
4                               Employed full-time           United States   
5                               Employed full-time                 Ukraine   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   
3                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
5                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                               UndergradMajor  \
Respondent                                                      
1                                                         NaN   
2                                                         NaN   
3                               Web development or web design   
4           Computer science, computer engineering, or sof...   
5           Computer science, computer engineering, or sof...   

                                                     EduOther  \
Respondent                                                      
1           Taught yourself a new language, framework, or ...   
2           Taken an online course in programming or softw...   
3           Taught yourself a new language, framework, or ...   
4           Taken an online course in programming or softw...   
5           Taken an online course in programming or softw...   

                             OrgSize  \
Respondent                             
1                                NaN   
2                                NaN   
3               100 to 499 employees   
4               100 to 499 employees   
5           10,000 or more employees   

                                                      DevType YearsCode  \
Respondent                                                                
1                                                         NaN         4   
2           Developer, desktop or enterprise applications;...       NaN   
3           Designer;Developer, back-end;Developer, front-...         3   
4                                       Developer, full-stack         3   
5           Academic re

## Getting a sense of the data

There are different way to know about the data.

In [142]:
df.shape

(88883, 84)

So, we can see that this CSV has 

1. Rows: 88,883
2. Columns: 84

When we did a head on the dataframe, we can only see 20 columsn, because by default it gets trimmed.

Even with this, I am not able to view all the columns.

So, I will confifure the head to show all.

In [143]:
pd.set_option('display.max_columns', 84)
# df.set_index('ResponseId', inplace=True) # adding the index to the df
df.head()

MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code      Yes   
2                      I am a student who is learning to code       No   
3           I am not primarily a developer, but I write co...      Yes   
4                              I am a developer by profession       No   
5                              I am a developer by profession      Yes   

                           OpenSourcer  \
Respondent                               
1                                Never   
2              Less than once per year   
3                                Never   
4                                Never   
5           Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   
5           OSS is, on average, of HIGHER quality than pro...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   
3                               Employed full-time                Thailand   
4                               Employed full-time           United States   
5                               Employed full-time                 Ukraine   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   
3                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
5                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                               UndergradMajor  \
Respondent                                                      
1                                                         NaN   
2                                                         NaN   
3                               Web development or web design   
4           Computer science, computer engineering, or sof...   
5           Computer science, computer engineering, or sof...   

                                                     EduOther  \
Respondent                                                      
1           Taught yourself a new language, framework, or ...   
2           Taken an online course in programming or softw...   
3           Taught yourself a new language, framework, or ...   
4           Taken an online course in programming or softw...   
5           Taken an online course in programming or softw...   

                             OrgSize  \
Respondent                             
1                                NaN   
2                                NaN   
3               100 to 499 employees   
4               100 to 499 employees   
5           10,000 or more employees   

                                                      DevType YearsCode  \
Respondent                                                                
1                                                         NaN         4   
2           Developer, desktop or enterprise applications;...       NaN   
3           Designer;Developer, back-end;Developer, front-...         3   
4                                       Developer, full-stack         3   
5           Academic re

Great, now I am able to view the entire data.

It's time to load the schema and see the questions.

So, I will load the other CSV file that I for in the zip file.

In [144]:
schema_df = pd.read_csv('../../data/stackoverflow/survey_results_schema.csv')
schema_df.sort_index(inplace=True)
schema_df.head()

Column                                       QuestionText
0   Respondent  Randomized respondent ID number (not in order ...
1   MainBranch  Which of the following options best describes ...
2     Hobbyist                            Do you code as a hobby?
3  OpenSourcer        How often do you contribute to open source?
4   OpenSource  How do you feel about the quality of open sour...

I am able to load the data. However, I can see that the question has HTML as well.

# Checking iloc and loc

In [145]:
schema_df.iloc[[1, 2, 3], 1]

1    Which of the following options best describes ...
2                              Do you code as a hobby?
3          How often do you contribute to open source?
Name: QuestionText, dtype: object

Looking at the iloc feature of DataFrames.

In [146]:
df.loc[1:4, "Hobbyist": "Student"]

Hobbyist              OpenSourcer  \
Respondent                                     
1               Yes                    Never   
2                No  Less than once per year   
3               Yes                    Never   
4                No                    Never   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   
3                               Employed full-time                Thailand   
4                               Employed full-time           United States   

                   Student  
Respondent                  
1                       No  
2           Yes, full-time  
3                       No  
4                       No

# Analysing data

In [147]:
df.head()

MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code      Yes   
2                      I am a student who is learning to code       No   
3           I am not primarily a developer, but I write co...      Yes   
4                              I am a developer by profession       No   
5                              I am a developer by profession      Yes   

                           OpenSourcer  \
Respondent                               
1                                Never   
2              Less than once per year   
3                                Never   
4                                Never   
5           Once a month or more often   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   
3           The quality of OSS and closed source software ...   
4           The quality of OSS and closed source software ...   
5           OSS is, on average, of HIGHER quality than pro...   

                                        Employment                 Country  \
Respondent                                                                   
1           Not employed, and not looking for work          United Kingdom   
2               Not employed, but looking for work  Bosnia and Herzegovina   
3                               Employed full-time                Thailand   
4                               Employed full-time           United States   
5                               Employed full-time                 Ukraine   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   
3                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
4                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
5                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                               UndergradMajor  \
Respondent                                                      
1                                                         NaN   
2                                                         NaN   
3                               Web development or web design   
4           Computer science, computer engineering, or sof...   
5           Computer science, computer engineering, or sof...   

                                                     EduOther  \
Respondent                                                      
1           Taught yourself a new language, framework, or ...   
2           Taken an online course in programming or softw...   
3           Taught yourself a new language, framework, or ...   
4           Taken an online course in programming or softw...   
5           Taken an online course in programming or softw...   

                             OrgSize  \
Respondent                             
1                                NaN   
2                                NaN   
3               100 to 499 employees   
4               100 to 499 employees   
5           10,000 or more employees   

                                                      DevType YearsCode  \
Respondent                                                                
1                                                         NaN         4   
2           Developer, desktop or enterprise applications;...       NaN   
3           Designer;Developer, back-end;Developer, front-...         3   
4                                       Developer, full-stack         3   
5           Academic re

In [148]:
country_count = df['Country'].value_counts(ascending = False)
country_count.head(10)

Country
United States     20949
India              9061
Germany            5866
United Kingdom     5737
Canada             3395
France             2391
Brazil             1948
Poland             1922
Australia          1903
Netherlands        1852
Name: count, dtype: int64

In [149]:
countries_lt_1500 = country_count[country_count<1500]
countries_lt_1500

Country
Sweden               1274
Switzerland           978
Israel                952
Turkey                949
Pakistan              923
                     ... 
Tonga                   1
Timor-Leste             1
North Korea             1
Brunei Darussalam       1
Chad                    1
Name: count, Length: 166, dtype: int64

In [150]:
def clean_country(value):
    if (value in countries_lt_1500):
        return 'Others'
    else:
        return value

In [151]:
df['Country'] = df['Country'].apply(clean_country)
df['Country'].unique()

array(['United Kingdom', 'Others', 'United States', 'Canada', 'India',
       'Germany', 'Australia', 'Russian Federation', 'Brazil', 'Spain',
       'Poland', 'France', 'Netherlands', 'Italy', nan], dtype=object)

In [156]:
df['Country'].value_counts(ascending=False)

Country
Others                28853
United States         20949
India                  9061
Germany                5866
United Kingdom         5737
Canada                 3395
France                 2391
Brazil                 1948
Poland                 1922
Australia              1903
Netherlands            1852
Russian Federation     1694
Spain                  1604
Italy                  1576
Name: count, dtype: int64

In [152]:
test = pd.DataFrame([
    {'firstName': 'Jhon', 'lastName': 'Wick', 'gender':'male'},
    {'firstName': 'Jack', 'lastName': 'Sparrow', 'gender':'male'},
    {'firstName': 'Maria', 'lastName': 'Sanchez', 'gender':'female'},
    {'firstName': 'Peter', 'lastName': 'Parker', 'gender':'male'},
])

test.head()

firstName lastName  gender
0      Jhon     Wick    male
1      Jack  Sparrow    male
2     Maria  Sanchez  female
3     Peter   Parker    male

In [153]:
test.loc[test['gender'] != 'female']

firstName lastName gender
0      Jhon     Wick   male
1      Jack  Sparrow   male
3     Peter   Parker   male

In [154]:
newT = (test['gender'] == 'male') & (test['firstName'] == 'Jhon')
result = test[~newT]
result

firstName lastName  gender
1      Jack  Sparrow    male
2     Maria  Sanchez  female
3     Peter   Parker    male

Was looking at some filtering techniques

# Trying to get users with High salary

In [165]:
df.head(2)

MainBranch Hobbyist  \
Respondent                                                    
1           I am a student who is learning to code      Yes   
2           I am a student who is learning to code       No   

                        OpenSourcer  \
Respondent                            
1                             Never   
2           Less than once per year   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
2           The quality of OSS and closed source software ...   

                                        Employment         Country  \
Respondent                                                           
1           Not employed, and not looking for work  United Kingdom   
2               Not employed, but looking for work          Others   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
2           Yes, full-time  Secondary school (e.g. American high school, G...   

           UndergradMajor                                           EduOther  \
Respondent                                                                     
1                     NaN  Taught yourself a new language, framework, or ...   
2                     NaN  Taken an online course in programming or softw...   

           OrgSize                                            DevType  \
Respondent                                                              
1              NaN                                                NaN   
2              NaN  Developer, desktop or enterprise applications;...   

           YearsCode Age1stCode YearsCodePro CareerSat JobSat MgrIdiot  \
Respondent                                                               
1                  4         10          NaN       NaN    NaN      NaN   
2                NaN         17          NaN       NaN    NaN      NaN   

           MgrMoney MgrWant                          JobSeek  \
Respondent                                                     
1               NaN     NaN                              NaN   
2               NaN     NaN  I am actively looking for a job   

                    LastHireDate LastInt FizzBuzz  \
Respondent                                          
1                            NaN     NaN      NaN   
2           I've never had a job     NaN      NaN   

                                                   JobFactors  \
Respondent                                                      
1                                                         NaN   
2           Financial performance or funding status of the...   

                                                 ResumeUpdate CurrencySymbol  \
Respondent                                                                     
1                                                         NaN            NaN   
2           Something else changed (education, award, medi...            NaN   

           CurrencyDesc  CompTotal CompFreq  ConvertedComp  WorkWeekHrs  \
Respondent                                                                
1                   NaN        NaN      NaN            NaN          NaN   
2                   NaN        NaN      NaN            NaN          NaN   

           WorkPlan WorkChallenge WorkRemote WorkLoc ImpSyn CodeRev  \
Respondent                                                            
1               NaN           NaN        NaN     NaN    NaN     NaN   
2               NaN           NaN        NaN     NaN    NaN     NaN   

            CodeRevHrs UnitTests PurchaseHow PurchaseWhat  \
Respondent                                                  
1                  NaN       NaN         NaN          NaN   
2                  NaN       Na

In [171]:
cols = ["Country", "EdLevel", "UndergradMajor", "OrgSize", "LanguageWorkedWith", "ConvertedComp"]
high_salary = (df["ConvertedComp"] > 70000)
df.loc[high_salary, cols]

Country                                            EdLevel  \
Respondent                                                                      
6                   Canada           Bachelor’s degree (BA, BS, B.Eng., etc.)   
9                   Others  Some college/university study without earning ...   
13           United States        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
16          United Kingdom        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
22           United States  Some college/university study without earning ...   
...                    ...                                                ...   
88876        United States           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88877        United States           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88878        United States           Bachelor’s degree (BA, BS, B.Eng., etc.)   
88879               Others        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
88882          Netherlands        Master’s degree (MA, MS, M.Eng., MBA, etc.)   

                                               UndergradMajor  \
Respondent                                                      
6                                   Mathematics or statistics   
9           Computer science, computer engineering, or sof...   
13          Computer science, computer engineering, or sof...   
16                                                        NaN   
22                                                        NaN   
...                                                       ...   
88876       Computer science, computer engineering, or sof...   
88877       Computer science, computer engineering, or sof...   
88878       Computer science, computer engineering, or sof...   
88879       Computer science, computer engineering, or sof...   
88882       Computer science, computer engineering, or sof...   

                             OrgSize  \
Respondent                             
6                                NaN   
9                 10 to 19 employees   
13                10 to 19 employees   
16              100 to 499 employees   
22          10,000 or more employees   
...                              ...   
88876       10,000 or more employees   
88877           500 to 999 employees   
88878             20 to 99 employees   
88879             20 to 99 employees   
88882           100 to 499 employees   

                                           LanguageWorkedWith  ConvertedComp  
Respondent                                                                    
6                                                  Java;R;SQL       366420.0  
9           Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;P...        95179.0  
13          Bash/Shell/PowerShell;HTML/CSS;JavaScript;PHP;...        90000.0  
16          Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;T...       455352.0  
22          Bash/Shell/PowerShell;C++;HTML/CSS;JavaScript;...       103000.0  
...                                                       ...            ...  
88876       Bash/Shell/PowerShell;C#;HTML/CSS;Java;Python;...       180000.0  
88877       Bash/Shell/PowerShell;C;Clojure;HTML/CSS;Java;...      2000000.0  
88878                    HTML/CSS;JavaScript;Scala;TypeScript       130000.0  
88879                        Bash/Shell/PowerShell;C++;Python        82488.0  
88882                  C#;HTML/CSS;Java;JavaScript;PHP;Python       588012.0  

[22289 rows x 6 columns]

In [174]:
interested_countries = df['Country'].isin(['United State', 'India', 'Germany', 'United Kingdom', 'Canada'])
df[interested_countries].shape

(24059, 84)

In [176]:
df[interested_countries].head(10)

MainBranch Hobbyist  \
Respondent                                                               
1                      I am a student who is learning to code      Yes   
6           I am not primarily a developer, but I write co...      Yes   
8                                 I code primarily as a hobby      Yes   
10                             I am a developer by profession      Yes   
12                     I am a student who is learning to code       No   
14                             I am a developer by profession      Yes   
15                     I am a student who is learning to code      Yes   
16                             I am a developer by profession      Yes   
21                                I code primarily as a hobby      Yes   
31                     I am a student who is learning to code       No   

                           OpenSourcer  \
Respondent                               
1                                Never   
6                                Never   
8              Less than once per year   
10          Once a month or more often   
12                               Never   
14             Less than once per year   
15                               Never   
16                               Never   
21          Once a month or more often   
31                               Never   

                                                   OpenSource  \
Respondent                                                      
1           The quality of OSS and closed source software ...   
6           The quality of OSS and closed source software ...   
8           OSS is, on average, of HIGHER quality than pro...   
10          OSS is, on average, of HIGHER quality than pro...   
12          OSS is, on average, of HIGHER quality than pro...   
14          The quality of OSS and closed source software ...   
15          OSS is, on average, of HIGHER quality than pro...   
16          The quality of OSS and closed source software ...   
21          OSS is, on average, of HIGHER quality than pro...   
31                                                        NaN   

                                        Employment         Country  \
Respondent                                                           
1           Not employed, and not looking for work  United Kingdom   
6                               Employed full-time          Canada   
8               Not employed, but looking for work           India   
10                              Employed full-time           India   
12                              Employed part-time          Canada   
14                              Employed full-time         Germany   
15              Not employed, but looking for work           India   
16                              Employed full-time  United Kingdom   
21                                             NaN          Canada   
31          Not employed, and not looking for work          Canada   

                   Student                                            EdLevel  \
Respondent                                                                      
1                       No                          Primary/elementary school   
6                       No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
8                      NaN           Bachelor’s degree (BA, BS, B.Eng., etc.)   
10                      No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
12          Yes, full-time  Some college/university study without earning ...   
14                      No          Other doctoral degree (Ph.D, Ed.D., etc.)   
15          Yes, full-time  Secondary school (e.g. American high school, G...   
16                      No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
21                     NaN                                                NaN   
31                      No                          Primary/elementary school   

                                               UndergradMajor 